File: github_code/ford-fulkerson/java/40.java
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.Map;
import java.util.Queue;
import java.util.Set;

public class FordFulkerson {

	public int maxFlow(int source, int sink, int cap[][]) {

		int result = 0;
		Map<Integer, Integer> parent = new HashMap<>();
		while (BFS(cap, source, sink, parent)) {
			int flow = Integer.MAX_VALUE;
			int v = sink;
			while (v != source) {
				int u = parent.get(v);
				flow = Math.min(flow, cap[u][v]);
				v = u;
			}
			result += flow;
			v=sink;
			while (v != source) {
				int u = parent.get(v);
				cap[u][v] -= flow;
				cap[v][u] += flow;
				v = u;
			}
		}
		return result;
	}

	public boolean BFS(int cap[][], int source, int sink, Map<Integer, Integer> parent) {

		Set<Integer> visited = new HashSet<>();
		Queue<Integer> queue = new LinkedList<>();
		visited.add(source);
		queue.add(source);
		boolean find = false;
		while (!queue.isEmpty()) {
			int u = queue.poll();
			for (int v = 0; v < cap.length; v++) {
				if (!visited.contains(v) && cap[u][v] > 0) {
					parent.put(v, u);
					visited.add(v);
					queue.add(v);
					if (v == sink) {
						find = true;
						break;
					}
				}
			}
		}
		return find;
	}

	public FordFulkerson() {
		int[][] capacity = { { 0, 3, 0, 3, 0, 0, 0 }, { 0, 0, 4, 0, 0, 0, 0 }, { 3, 0, 0, 1, 2, 0, 0 },
				{ 0, 0, 0, 0, 2, 6, 0 }, { 0, 1, 0, 0, 0, 0, 1 }, { 0, 0, 0, 0, 0, 0, 9 }, { 0, 0, 0, 0, 0, 0, 0 } };
		System.out.println(maxFlow(0, 6, capacity));
	}

	public static void main(String[] args) {
		new FordFulkerson();
	}
}

File: github_code/ford-fulkerson/cpp/40.c
#include "stdafx.h"
#include "iostream"
#include "SelfStudyFunctions.h"
#include "list"
#include <limits>
#include <map>
#include <limits>
#include <string>

FordFulkerson::FordFulkerson() {}

FordFulkerson::FordFulkerson(int s) {
	this->graph.resize(s);
	this->size = s;
	for (int idx = 0; idx < size; idx++) {
		graph[idx].resize(s, 0);
	}
}

bool FordFulkerson::getBFSPath(int start, int end, std::vector<int> &parent, std::vector<std::vector<int>> rGraph) {
	std::list<int> queue;
	queue.push_back(start);
	std::vector<bool> visited(this->size, false);
	parent[start] = start;
	bool pathFound = false;
	while (queue.size() > 0 && !pathFound) {
		int node = queue.front();
		visited[node] = true;
		queue.pop_front();
		for (int i = 0; i < this->size; i++) {
			if (rGraph[node][i] > 0 && !visited[i]) {
				parent[i] = node;
				queue.push_back(i);
				if (i == end) {
					pathFound = true;
					break;
				}
			}
		}
	}
	return pathFound;
}

int FordFulkerson::getMaxFlow(int start, int end) {
	std::vector<std::vector<int>> rGraph = this->graph;
	std::vector<int> parent(this->size, -1);
	int maxFlow = 0;
	while (getBFSPath(start, end, parent, rGraph)) {
		int child = end;
		int par = -1;
		int minFlowForPath = std::numeric_limits<int>::max();
		//Loop to find min flow
		while (par != start) {
			par = parent[child];
			minFlowForPath = (rGraph[par][child] < minFlowForPath) ? rGraph[par][child] : minFlowForPath;
			child = par;
		}
		//Loop to update residual graph
		par = -1;
		child = end;
		while (par != start) {
			par = parent[child];
			rGraph[par][child] -= minFlowForPath;
			rGraph[child][par] += minFlowForPath;
			child = par;
		}
		maxFlow += minFlowForPath;
	}
	return maxFlow;
}

void FordFulkerson::testGetMaxFlow() {
	FordFulkerson ff(6);
	ff.graph[0][1] = 16;
	ff.graph[1][3] = 12;
	ff.graph[3][5] = 20;
	ff.graph[0][2] = 13; 
	ff.graph[2][4] = 14; 
	ff.graph[4][5] = 4; 
	ff.graph[1][2] = 10; 
	ff.graph[2][1] = 4;
	ff.graph[3][2] = 9;
	ff.graph[4][3] = 7;
	std::cout << "Max flow = " << ff.getMaxFlow(0, 5);
}